### PR TITLE
feature/provisioner_rework

### LINKED ISSUE #7597: Autoregistration with a provisionner doesn't work as expected
**Is your feature request related to a problem? Please describe.**
If you configure a connection profile with:
  * autoregistration
  * a provisionner which apply to role assigned to 802.1X devices doing autoregistration

The provisionner will trigger a delayed security event on 802.1X devices. If they are not compliant, PacketFence will:
  * switch their status from registered to pending
  * reevaluate access

It means devices will go on registration VLAN and if they go on captive portal, provisionner will display actions to become compliant.

**However**, if supplicant is still running on 802.1X devices, they could autoregister again (without any action). In fact, devices status will switch from pending to registered and security event will not be triggered due to grace time. 
It means that they will gain access to production network even if they are not compliant.

**Describe the solution you'd like**
As discussed with @fdurand, we must enforce the fact that devices are not compliant using a security event and release it once they are compliant.

**Describe alternatives you've considered**
I don't see other solutions at the moment.

**Additional context**
Example of connection profile and provisionner used to replicate:

```
# profiles.conf, dot1x_eap_peap_user_auth will assign user_employee role to devices
[provisionner_and_intune]
filter=connection_type:Ethernet-EAP
advanced_filter=
description=provisionner_and_intune
locale=
provisioners=intune
autoregister=enabled
sources=dot1x_eap_peap_user_auth

# provisioning.conf (only contains relevant information)
[intune]
autoregister=disabled
category=user_employee
type=intune
enforce=enabled
apply_role=disabled
sync_pid=disabled
[..]
```

### PR DESCRIPTION
**Pull request by jrouzierinverse**
# Description
Provide greater flexibility to provisioning workflow by adding filtering and actions.

# Impacts
Provisioning workflow

# Issue
fixes #7597

# Delete branch after merge
YES

# Checklist
(REQUIRED) - [yes, no or n/a]
- [ ] Document the feature
- [*] Add OpenAPI specification
- [ ] Add unit tests
- [ ] Add acceptance tests (TestLink)

# NEWS file entries

## Enhancements
* Provide greater flexibility to provisioning workflow by adding filtering and actions.

### CODE REVIEW

File: docs/api/spec/components/schemas/configconnectionprofile.yaml [hunk 1]
```diff
@@ -141,7 +141,7 @@ ConfigConnectionProfile:
       type: string
     network_logoff:
       default: ~
-      description: This allows users to access the network logoff page (http://packetfence.packetfence.org/networklogoff)
+      description: This allows users to access the network logoff page (http://172-105-101-170.ip.linodeusercontent.com/networklogoff)
```
Comment (satkunas) [inline docs/api/spec/components/schemas/configconnectionprofile.yaml]: We should avoid publishing this in the spec. Can we use PCRE post-generation to replace public IPv4/hostname with a placeholder?

File: addons/dev-helpers/bin/provisioning_filters_flatten.pl [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+    <$fh>
+};
+
+my $json = decode_json($data);
+my @fields;
+flatten("", $json, \@fields);
+@fields = sort @fields;
+if (defined $entry) {
+    my $ini = pf::IniFiles->new(-file => $provisioning_filters_meta_config_default_file) or die "";
+    use Data::Dumper;
```
Comment (fdurand) [inline addons/dev-helpers/bin/provisioning_filters_flatten.pl]: Already on top

File: addons/dev-helpers/bin/provisioning_filters_flatten.pl [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+    print Dumper($ini);
+    $ini->AddSection($entry);
+    $ini->delval($entry, 'fields');
+    $ini->newval($entry, 'fields', @fields);
+    #$ini->push($entry, @fields);
```
Comment (fdurand) [inline addons/dev-helpers/bin/provisioning_filters_flatten.pl]: remove

File: docs/api/spec/components/schemas/configdomain.yaml [hunk 4]
```diff
@@ -38,7 +38,7 @@ ConfigDomain:
       description: Password of the machine account to be added to Active Directory.
       type: string
     ntlm_auth_host:
-      default: 127.0.0.1
+      default: 100.64.0.1
```
Comment (fdurand) [inline docs/api/spec/components/schemas/configdomain.yaml]: wrong PR i think

File: docs/api/spec/components/schemas/configsourcesubtypesponsoremail.yaml [hunk 5]
```diff
@@ -183,7 +183,7 @@ ConfigSourceSubTypeSponsoremail:
       description: Id
       type: string
     lang:
-      default: en_US
+      default: fr_FR
```
Comment (fdurand) [inline docs/api/spec/components/schemas/configsourcesubtypesponsoremail.yaml]: Why ?

File: docs/api/spec/openapi.json [hunk 6]
```diff
@@ -5578,7 +5578,7 @@
                   "type" : "string"
                },
                "ntlm_auth_host" : {
-                  "default" : "127.0.0.1",
+                  "default" : "100.64.0.1",
```
Comment (fdurand) [inline docs/api/spec/openapi.json]: Should be in devel instead

File: html/pfappserver/lib/pfappserver/Form/Config/Provisioning.pm [hunk 7]
```diff
@@ -123,6 +125,18 @@ has_field 'oses' =>
...
+has_field 'rules' =>
+  (
+   type => 'Select',
+   multiple => 1,
+   label => 'Roles',
+   options_method => \&options_rules,
+   element_class => ['chzn-deselect'],
+   element_attr => {'data-placeholder' => 'Click to add a rule'},
+   tags => { after_element => \&help,
+             help => 'Nodes with the selected roles will be affected' },
```
Comment (fdurand) [inline html/pfappserver/lib/pfappserver/Form/Config/Provisioning.pm]: Wrong help.

File: lib/pf/ConfigStore/ProvisioningFilters.pm [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+
+__PACKAGE__->meta->make_immutable unless $ENV{"PF_SKIP_MAKE_IMMUTABLE"};
+
+=head1 AUTHOR
+
+Inverse inc. <info@inverse.ca>
+
+=head1 COPYRIGHT
+
+Copyright (C) 2005-2023 Inverse inc.
```
Comment (fdurand) [inline lib/pf/ConfigStore/ProvisioningFilters.pm]: 2024

File: lib/pf/UnifiedApi/Controller/Config/FilterEngines/ProvisioningFilters.pm [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+    return ($cs->is_section_in_import($id)) ? $self->json_true : $self->SUPER::is_sortable($cs, $id, $item);
+}
+ 
+=head1 AUTHOR
+
+Inverse inc. <info@inverse.ca>
+
+=head1 COPYRIGHT
+
+Copyright (C) 2005-2023 Inverse inc.
```
Comment (fdurand) [inline lib/pf/UnifiedApi/Controller/Config/FilterEngines/ProvisioningFilters.pm]: 2024

**Pull request merged**